proposed/2023/release-staging-and-deprecation.md

```diff
@@ -0,0 +1,199 @@
+A new NuGet CLI command `stage` will be added, which will allow a user to stage a package to nuget, specifying a `group` parameter.
+ Staged packages will be uploaded to NuGet and will be validated during the staging process, so that the staging fails
+ if validation fails for a package in a group. After uploading multiple packages, the user user can then call `nuget push --group [group]`
+ command to request all the packages from the earlier group to become availble on the server as `published`.
+Note, that the `group` information will only be stored on the server and has nothing to do with how a package was built.
+
+Groups are scoped to an owner. So a package published by `Owner 1` into `Group 1` has nothing to do with another package published
+ to `Group 1` by `Owner 2`, as these should be treated as two separate groups.
+
+Groups can expand over time, by publishing new packages to the same group in the future. There is no concept of a "deprecated group"
+ as group by itself is stateless. It's just a collection of packages. A package can belong only to a single group.
+ This group expansion functionality will be handy for the error situation described above, where during staging a package staging fails.
+ This will result in only some packages being staged. The user will need to address the issues related to the failed package
+ and stage the remaining packages using the same command. Note, that if a package is stages, the operation should be idempotent,
+ and a request to stage the same package again should only result in a warning message (package is already staged) and the package shouldn't be uploaded again.
+
+Groups can later be used in the future when it's time to deprecate a set of packages. The package author can request to
+ deprecate a `group` by calling a single command on the nuget CLI and all the packages that belong to that group will be marked as `deprecated`.
+
+With this information in place, new set of NuGet APIs / extensions will be needed to allow a user to run the following actions:
+- **Stage** a package for later publishing
+- **Publish an earlier staged release**
+- **Deprecate a group** using the group id 
+- **Discard a staged release** using the group id
+
+Note, that the user is required to be the owner / co-owner of the packages which are in the group,
+otherwise the request to make any modification to a non-owned package will fail.
+
+#### Package staging
+When a package is staged, the id of the package is reserved at that point, and no other user can introduce a package with the same name.
+
+Also, to avoid abuse, only a limited number of packages can be staged within a single group. While this can potentially be defined as a configurable value per organization / owner, there is no such requirement at the moment. Hence, this can be set to 1000 packages for now or not defined at all. Will leave this for potentially future consideration.
```

Comment:
Does 1000 work for .NET releases? I think this is an important bound to understand sooner rather than later since impacts the viability of a "stage publish SLO" (how long it takes to publish up to X packages quickly)

Reply:
I checked and for .NET 8 release we had a bit less than 700 packages. So 1000 is a reasonable default cap, I think.